test_fms/monin_obukhov/test_monin_obukhov.F90

```diff
@@ -17,253 +17,517 @@
+                            & 0.00262353784027441_kr, -0.000570314880866852_kr]
+      profile_input%q_star = [0.000110861442197537_kr, 9.44983279664197e-05_kr, 4.17643828631936e-05_kr, &
+                            & 0.000133135421415819_kr, 9.36317815993945e-06_kr]
+      profile_input%avail = [.true., .true., .true., .true., .true.]
+
+      diff_input%z      = 19.9982554527751_kr
+      diff_input%u_star = 0.129638955971075_kr
+      diff_input%b_star = 0.000991799765557209_kr
+    end subroutine
+
+    !< Read answer keys from input.nml
+    subroutine read_answers
+      integer :: io, ierr
+
+      read (input_nml_file, nml=metaparams_nml, iostat=io)
+      ierr = check_nml_error(io, "metaparams_nml")
+
+      allocate(drag_answers(n_answers+1))
+      allocate(stable_mix_answers(n_answers+1))
+      allocate(diff_answers(n_answers+1))
+      allocate(profile_answers(n_answers+1))
+
+      if (n_answers.gt.0) then
+        read (input_nml_file, nml=answers_nml, iostat=io)
+        ierr = check_nml_error(io, "answers_nml")
+      endif
+    end subroutine
+
+    !> Store existing answer keys, as well as the answers just calculated, in an
+    !> output file.
+    subroutine write_answers
+      character(:), allocatable :: filename
+      integer :: fh
+
+      filename = "OUT.r" // string(MO_TEST_KIND_) // ".nml"
+      print "(A)", "Writing newly generated answer key to " // filename
+
+      n_answers = n_answers + 1
+
+      open (newunit=fh, file=filename)
+      write (fh, nml=metaparams_nml)
+      write (fh, nml=answers_nml)
+      close (fh)
+    end subroutine
+
+    !> Calculate all answers
+    subroutine calc_answers
+      call calc_answers_drag
+      call calc_answers_stable_mix
+      call calc_answers_diff
+      call calc_answers_profile
+    end subroutine
+
+    !> Calculate 1D answers for mo_drag, and assert that all 2D answers must agree
+    !> with the corresponding 1D answers
+    subroutine calc_answers_drag
+      real(kr), dimension(n_1d) :: drag_m_1d, drag_t_1d, drag_q_1d, u_star_1d, b_star_1d
+      real(kr), dimension(n_1d, n_1d) :: drag_m_2d, drag_t_2d, drag_q_2d, u_star_2d, b_star_2d
+
+      drag_m_1d = 0._kr
+      drag_t_1d = 0._kr
+      drag_q_1d = 0._kr
+      u_star_1d = 0._kr
+      b_star_1d = 0._kr
+
+      drag_m_2d = 0._kr
+      drag_t_2d = 0._kr
+      drag_q_2d = 0._kr
+      u_star_2d = 0._kr
+      b_star_2d = 0._kr
+
+      associate (in => drag_input)
+        call mo_drag(in%pt, in%pt0, in%z, in%z0, in%zt, in%zq, in%speed, &
+                   & drag_m_1d, drag_t_1d, drag_q_1d, u_star_1d, b_star_1d, drag_input%avail)
+
+        call mo_drag(ARR_2D_(in%pt), ARR_2D_(in%pt0), ARR_2D_(in%z), ARR_2D_(in%z0), &
+                   & ARR_2D_(in%zt), ARR_2D_(in%zq), ARR_2D_(in%speed), &
+                   & drag_m_2d, drag_t_2d, drag_q_2d, u_star_2d, b_star_2d)
+      end associate
+
+      associate(ans => drag_answers(n_answers+1))
+        ans%drag_m = INT_(drag_m_1d)
```

Comment:
why are the answers converted to itnegers?

Reply:
Reference answers are loaded from a namelist file so that it's easy to add additional answer keys for new compilers or platforms. With reals, the Fortran standard makes no guarantee that `x.eq.y` is true, where `y` is the result of storing `x` in a namelist file and reading it back. Integers have unambiguous string representations however, so `x.eq.y` will be true if `y` is the result of `transfer()`-ing `x` to an integer, storing that integer in a namelist file, reading it back, and `transfer()`-ing it back to a real.

Reply:
I attempted that approach at first. It would be more elegant for sure if I could spend a semester studying the math. But given the timeline, I decided that a lazy reference answer approach would be more expedient.